src/net.h

```diff
@@ -1257,6 +1276,40 @@ class CConnman
      */
     std::vector<CService> m_onion_binds;
 
+    /**
+     * RAII helper to atomically create a copy of `vNodes` and add a reference
+     * to each of the nodes. The nodes are released when this object is destroyed.
+     */
+    class NodesSnapshot
```

Comment:
I'm guessing you've defined this class within `CConnman` to enable access to the private members. You could instead define the class in `net.cpp` as a standalone by making `vNodes` and `cs_vNodes` protected & declaring `friend class NodesSnapshot` here. Seems nice to keep this mechanism that's only used internally out of the header file.

diff 👇🏽
```diff 
diff --git a/src/net.cpp b/src/net.cpp
index 5fd1bf754d..6040ee86be 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -110,6 +110,39 @@ std::map<CNetAddr, LocalServiceInfo> mapLocalHost GUARDED_BY(cs_mapLocalHost);
 static bool vfLimited[NET_MAX] GUARDED_BY(cs_mapLocalHost) = {};
 std::string strSubVersion;

+/**
+ * RAII helper to atomically create a copy of `vNodes` and add a reference
+ * to each of the nodes. The nodes are released when this object is destroyed.
+ */
+class NodesSnapshot
+{
+    public:
+        explicit NodesSnapshot(const CConnman& connman) : m_connman{connman}, m_nodes_copy{m_connman.vNodes}
+        {
+            LOCK(m_connman.cs_vNodes);
+            for (auto& node : m_nodes_copy) {
+                node->AddRef();
+            }
+        }
+
+        ~NodesSnapshot()
+        {
+            LOCK(m_connman.cs_vNodes);
+            for (auto& node : m_nodes_copy) {
+                node->Release();
+            }
+        }
+
+        const std::vector<CNode*>& Nodes() const
+        {
+            return m_nodes_copy;
+        }
+
+    private:
+        const CConnman& m_connman;
+        const CConnman& m_connman;
+        const std::vector<CNode*> m_nodes_copy;
+};
+
 void CConnman::AddAddrFetch(const std::string& strDest)
 {
     LOCK(m_addr_fetches_mutex);
diff --git a/src/net.h b/src/net.h
index 846c4c0ece..f39c46c166 100644
--- a/src/net.h
+++ b/src/net.h
@@ -1021,6 +1021,10 @@ public:
     /** Return true if we should disconnect the peer for failing an inactivity check. */
     bool ShouldRunInactivityChecks(const CNode& node, std::optional<int64_t> now=std::nullopt) const;

+protected:
+    std::vector<CNode*> vNodes GUARDED_BY(cs_vNodes);
+    mutable RecursiveMutex cs_vNodes;
+
 private:
     struct ListenSocket {
     public:
@@ -1153,9 +1157,7 @@ private:
     RecursiveMutex m_addr_fetches_mutex;
     std::vector<std::string> vAddedNodes GUARDED_BY(cs_vAddedNodes);
     mutable RecursiveMutex cs_vAddedNodes;
-    std::vector<CNode*> vNodes GUARDED_BY(cs_vNodes);
     std::list<CNode*> vNodesDisconnected;
-    mutable RecursiveMutex cs_vNodes;
     std::atomic<NodeId> nLastNodeId{0};
     unsigned int nPrevNodeCount{0};

@@ -1276,40 +1278,7 @@ private:
      */
     std::vector<CService> m_onion_binds;

-    /**
-     * RAII helper to atomically create a copy of `vNodes` and add a reference
-     * to each of the nodes. The nodes are released when this object is destroyed.
-     */
-    class NodesSnapshot
-    {
-    public:
-        explicit NodesSnapshot(const CConnman& connman) : m_connman{connman}
-        {
-            LOCK(m_connman.cs_vNodes);
-            m_nodes_copy = m_connman.vNodes;
-            for (auto& node : m_nodes_copy) {
-                node->AddRef();
-            }
-        }
-
-        ~NodesSnapshot()
-        {
-            LOCK(m_connman.cs_vNodes);
-            for (auto& node : m_nodes_copy) {
-                node->Release();
-            }
-        }
-
-        const std::vector<CNode*>& Nodes() const
-        {
-            return m_nodes_copy;
-        }
-
-    private:
-        const CConnman& m_connman;
-        std::vector<CNode*> m_nodes_copy;
-    };
-
+    friend class NodesSnapshot;
     friend struct CConnmanTest;
     friend struct ConnmanTestMsg;
 };
```

Reply:
I like this. A few notes:

- the diff as shown is corrupted (contains the `+        const CConnman& m_connman;` line twice) and so doesn't apply with `git apply`.
- `vNodes` and `cs_vNodes` don't need to change to be protected. Friend classes have access to private members.
- `NodesSnapshot` can be in the unnamed namespace since it's only needed in the net.cpp translation unit.

Reply:
The `m_nodes_copy{m_connman.vNodes}` in the initializer list is accessing `vNodes` without the `cs_vNodes` lock. You could update it to take the lock:

```diff
diff --git a/src/net.cpp b/src/net.cpp
index 6040ee86be..d39aa14b20 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -117,7 +117,9 @@ std::string strSubVersion;
 class NodesSnapshot
 {
     public:
-        explicit NodesSnapshot(const CConnman& connman) : m_connman{connman}, m_nodes_copy{m_connman.vNodes}
+        explicit NodesSnapshot(const CConnman& connman) :
+            m_connman{connman},
+            m_nodes_copy{WITH_LOCK(m_connman.cs_vNodes, return m_connman.vNodes;)}
         {
             LOCK(m_connman.cs_vNodes);
             for (auto& node : m_nodes_copy) {
```

But that would introduce a race condition if another thread deleted a node between the initializer list and the ctor being run.

I think better just to assign `m_nodes_copy` in the ctor.

Reply:
> I'm guessing you've defined this class within CConnman to enable access to the private members.

No, I did it to minimize the scope of the newly added class.

> Seems nice to keep this mechanism that's only used internally out of the header file.

Why? It is used internally by `CConnman` and is defined in its `private:` section. If defined in `net.cpp` that would broaden its scope unnecessary.

Reply:
> > Seems nice to keep this mechanism that's only used internally out of the header file.

> Why?

Because it's good to keep header files limited to what's required to be included into other translation units (as far as is possible). That's beneficial for both build time and code management reasons.

I think both ways are fine, and that eventually we'll want to virtualize both `CNode`'s and `CConnman`'s public interfaces so that they can be mocked out for testing, at which point the private methods could all be moved to the implementation file.

Reply:
> That's beneficial for both build time...

I do not think it is justified to put private class members as globals in the `.cpp` file in order to minimize the header file so that it compiles faster.

> ... and code management reasons.

Unnecessary broadened scope has an adverse effect on code management.

Reply:
If `NodesSnapshot` was defined as a standalone class in `net.cpp`, it could be added to the unnamed namespace so it would only be accessible within the same translation unit, but it would broaden the scope beyond `CConnman` to within the translation unit. 

My preference for it to be in the `.cpp` is because I find it easier to understand responsibilities & maneuver the files when the headers capture the way the classes interact with external callers. But I totally understand the point about broadening scope, and think this ultimately comes down to a preference.

I'm going to resolve this conversation, thanks for the feedback @vasild & @jnewbery :)

Reply:
@amitiuttarwar, maybe you would like the following pattern to be more widespread:

https://github.com/bitcoin/bitcoin/blob/1ed859e90e18384376e3a1ff0cb76f3e9ab11c2d/src/txrequest.h#L96-L101

Reply:
yeah, I'm def a fan of the pimpl / related patterns. Another example is how its used in `net_processing` since #20811.